### PR TITLE
in_tail: Avoid blocking on pending pipe

### PR DESCRIPTION
**Pull request by jsravn**
When paused under heavy load, it is possible for the pending pipe to
fill up and then block - causing the entire fluent-bit process to
block/hang. This is due to adding a pending signal for every file, instead
of just one single signal. Similarly, every inotify event can also fill the pending pipe.

Fix is to use nonblocking mode for the pending pipe, and clear it out after all pending files are processed.

### CODE REVIEW

**Comment by jsravn** [comment]
I missed another spot where pipe can fill up in the inotify code

```
#0  0x00007f1fbc083a3d in write () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00000000004901e3 in tail_signal_pending (ctx=0x7f1fba677140) at /tmp/fluent-bit-sky-build/plugins/in_tail/tail_signal.h:47
#2  0x000000000049040e in tail_fs_event (i_ins=0x7f1fba657000, config=0x7f1fba61c0c0, in_context=0x7f1fba677140)
    at /tmp/fluent-bit-sky-build/plugins/in_tail/tail_fs_inotify.c:110
#3  0x000000000046d5d8 in flb_input_collector_fd (fd=20, config=0x7f1fba61c0c0) at /tmp/fluent-bit-sky-build/src/flb_input.c:987
#4  0x0000000000473e57 in flb_engine_handle_event (config=0x7f1fba61c0c0, mask=1, fd=20) at /tmp/fluent-bit-sky-build/src/flb_engine.c:292
#5  flb_engine_start (config=0x7f1fba61c0c0) at /tmp/fluent-bit-sky-build/src/flb_engine.c:502
#6  0x000000000041c6f7 in main (argc=2, argv=0x7ffe2f3fb248) at /tmp/fluent-bit-sky-build/src/fluent-bit.c:729
```

I'll fix this shortly.

**Comment by jsravn** [comment]
Okay, pushed a commit which handles the inotify case. I'm pretty sure that's the last spot where it can block. Will test this out further in my own clusters.

**Comment by jsravn** [comment]
After testing for a couple days, haven't seen a hanging fluent-bit yet.

File: plugins/in_tail/tail.c [hunk 1]
```diff
@@ -106,6 +104,11 @@ static int in_tail_collect_pending(struct flb_input_instance *i_ins,
         }
     }
 
+    /* If no more active files, consume pending signal so we don't get called again. */
+    if (active == 0) {
+        tail_consume_pending(ctx);
+    }
```
Comment (jsravn) [inline plugins/in_tail/tail.c]: This change isn't necessary now that I made the pipe NONBLOCK. However, it should reduce the number of duplicate signals added to the pending pipe - so I think it's worth keeping.

**Comment by StevenACoffman** [comment]
I tested this in my cluster and it does improve reliability substantially. Nice fix!

**Comment by jsravn** [comment]
If anyone else wants to try it out, you can use my docker image: `skycirrus/fluent-bit-0.12.8:10`.

**Pull request merged**